src/book/02-projects/01-nubots/12-mathematics.mdx

```diff
@@ -0,0 +1,158 @@
+
+**Rotations** are defined by a 3x3 matrix. If we have a vector $\mathbf{r}^a$, that is the vector $r$ in the coordinate system $a$, the rotation
+
+$$
+\begin{bmatrix}
+\cos(\psi) & -\sin(\psi) & 0 \\
+\sin(\psi) & \cos(\psi) & 0 \\
+0 & 0 & 1 \end{bmatrix}
+$$
+
+will rotate $\mathbf{r}^a$ by $\psi$ around the coordinate axis $\mathbf{a_3}$.
+
+The orientation of the basis $b$ relative to $a$ can be described by three consecutive rotations about the main axis. There are 12 different orders to do these rotations in, and each triplet of rotated angles is called a set of **Euler angles**. In robotics, the triplet is **yaw**, **pitch** and **roll** (Z-Y-X).
+
+In the NUbots codebase, we denote a rotation matrix from space $b$ to space $a$ as `Rab`.
+
+There rotation matrices belong to the **special orthogonal group of order 3**. These all have the following properties:
+
+$$
+\mathbf{R}^{-1} = \mathbf{R}^T
+$$
+
+$$
+\text{det}(\mathbf{R}) = 1
+$$
+
+If we have a vector $\mathbf{r}$ (which goes from point C to point D) in a space $a$ and we want to rotate it from space $a$ to space $b$, we write:
+
+$$
+\mathbf{r}^b = \mathbf{R}^b_a \cdot \mathbf{r}^a
+$$
+
+This is written in the NUbots codebase as `rCDb = Rba * rCDa`.
+
+## Transformations
+
+A transformation matrix in three dimensions is a 4x4 matrix containing a **rotation** component and a **translation** component.
+
+If we have some rotation $\mathbf{R}_a^b$ and a translation component $\mathbf{r}^b$, then we have a transformation matrix that transforms from space $a$ to space $b$:
+
+$$
+\mathbf{H}_a^b = \begin{bmatrix}
+\mathbf{R}_a^b & \mathbf{r}_{AB}^b \\
+\mathbf{0} & 1
+\end{bmatrix}
+$$
+
+This is written in the NUbots codebase as `Hba`.
+
+The vector $\mathbf{r}_{AB}^b$ is the vector from B (the origin of space b) to A (the origin of space a) in b space, which would be written in the codebase as `rABb`.
+
+The reason for this is that if we apply the transform $H^b_a$ to some point $\rho^a$, then $\rho^a$ will first be rotated by $R_a^b$, giving $\rho^b$. $\rho^b$ will then be translated by some vector according to the transformation matrix. Since $\rho^b$ is now rotated in the $b$ coordinate system, the translation vector must also be in this $b$ coordinate system. The vector is from B to A because a point (0, 0, 0) in space $a$ should be moved to the origin of $a$ in $b$ space after transformation. Thus, the translation vector is $\mathbf{r}_{AB}^b$.
+
+Transformation matrices left multiply with vectors to get a vector in the new space.
+
+$$
+\mathbf{r}^b = \mathbf{H}^b_a \cdot \mathbf{r}^a
+$$
+
+This is written in the codebase as `rCDb = Hba * rCDa`, if $\mathbf{r}$ represents a vector from point D to point C.
+
+They also multiply together to get a new transformation matrix.
+
+$$
+\mathbf{H}_a^c = \mathbf{H}_b^c \cdot \mathbf{H}^b_a
+$$
+
+This is written in the codebase as `Hca = Hcb * Hba`.
+
```

Comment:
We should highlight the difference between points and vectors as they relate to homogeneous transformations.

A point is a vec4 `(x, y, z, 1)`. A point can be rotated and translated.
A vector is a vec4 `(x, y, z, 0)`. A vector can only be rotated.

Reply:
Do we have any notation for specifying a point? Like how we have rABc for vectors?

Reply:
Our vector notation came from a system where we didn't use homogenous transforms. In that system, there was no need to distinguish as everything was a vector from one point to another. So a vector from the origin of A to some point B is a point because it's referenced to the origin of the space. 99% of the time when we are using our vectors we treat them as points. So those matrices apply rotations and then translations.

Reply:
That sounds real messy.